<h3><b>Association Rule and Lift Analysis</b></h3>

A1. Proposal of Question: 
What are the products that, when combined with discounts, could lower customer churn? For instance, can we learn more about which products, if supplied at a deal with our services, can make us more popular with customers by studying a list of transactions?

A2. Objective: With some degree of confidence, the company's stakeholders will gain from knowing which customers are most likely to churn. This information will help them decide how to market improved services to customers who fit these criteria and have a history of dissatisfaction. This data analysis aims to showcase discounted products for company stakeholders to consider when coming up with consumer incentives and marketing promotions. We will work to make decision-makers more aware of certain feature combinations—items used in conjunction with telecom services—to lower the risk of churn among their clients.

B1. Theory Behind Market Basket Analysis:

Market basket analysis is one of the most critical methods for identifying item correlations. It operates by looking for product combinations that regularly appear together in transactions. This investigation aims to determine which telecom peripheral and ICT tool combinations people prefer and most frequently buy. To show the connections between these various things, we will attempt to pinpoint the ones that are most commonly purchased in pairs. We anticipate finding the ideal set of products to combine with our services and provide a discount. Preparing the dataset to find missing values is part of our analysis method. The Apriori method will be used to discover association rules. Then later, the lift, support, and confidence rules are verified.

B2. Transaction Example: 
The transactions in the provided dataset can be quickly identified. A longer list of twenty goods appears in the very first transaction, including:
* Wireless mouse made by Logitech
* Nonda HP 63 ink
* HP 65 ink 
* USB Type C cable
* 10-foot iPhone charger cable 
* USB C to USB adapter
* Creative Pebble 2.0 Speakers 
* HP 902XL ink
* Cosmetic Gel Micro Center
* 32GB Memory card 
* Universal Dust Cleaner
* 3pack 6ft Nylon Lightning Cable from YUNSONG
* Cooler pad (TopMate)
* USB-C cable for charging. Product by Apple
* Cloud Stinger headset by HyperX
* TONOR USB Gaming Microphone 
* Compressed Gas Dust-Off
* iPhone Charger (HOVAMP)
* SanDisk Ultra 128GB card
* FEEL2NICE 10ft Lightning cable
* Blue-light-blocking glasses from FEIYOLD

One client made simultaneous purchases of all twenty of these goods.

B3. Assumption for Market Basket Analysis (MBA):

Making decisions based on building association regulations is one of the MBA's presumptions. Dr. Susan Sivek says that these guidelines "are merely connections between "antecedent" and "consequent" items. Additionally, association rules only suggest co-occurrence and not causality " (Sivek, p. 1). The purchase of one will prompt the purchase of the others.

C1. Transforming the Dataset for Market Basket Analysis:

In [2]:
import pandas as pd

In [3]:
telecompany = pd.read_csv('teleco_market_basket.csv')


In [4]:
null_values= telecompany.isnull().sum()

In [5]:
telecompany.dropna(how = 'all', inplace = True)

In [6]:
telecompany.fillna(0, inplace = True)

In [7]:
telecompany.shape

(7501, 20)

In [8]:
#Convert dataset into list format for the use of Apriori algorithm

listed = []
for x in range(0, 7501):
 listed.append([str(telecompany.values[x, y]) for y in range(0, 20)])
telecompany_cleaned = pd.DataFrame(listed)

In [9]:
telecompany_cleaned.to_csv('prepared.csv')

In [10]:
listed[:1]

[['Logitech M510 Wireless mouse',
  'HP 63 Ink',
  'HP 65 ink',
  'nonda USB C to USB Adapter',
  '10ft iPHone Charger Cable',
  'HP 902XL ink',
  'Creative Pebble 2.0 Speakers',
  'Cleaning Gel Universal Dust Cleaner',
  'Micro Center 32GB Memory card',
  'YUNSONG 3pack 6ft Nylon Lightning Cable',
  'TopMate C5 Laptop Cooler pad',
  'Apple USB-C Charger cable',
  'HyperX Cloud Stinger Headset',
  'TONOR USB Gaming Microphone',
  'Dust-Off Compressed Gas 2 pack',
  '3A USB Type C Cable 3 pack 6FT',
  'HOVAMP iPhone charger',
  'SanDisk Ultra 128GB card',
  'FEEL2NICE 5 pack 10ft Lighning cable',
  'FEIYOLD Blue light Blocking Glasses']]

C2: Execution

In [11]:
# from the Apriori algorithm, produce association rules
from apyori import apriori
#on the dataset, practice the Apriori algorithm.
rules= apriori(listed, min_support = 0.0035, min_confidence = 0.22, min_lift = 3, min_length = 2)

In [12]:
# Review and produce rules
rules= list(rules)

In [13]:
print(len(rules))

84


In [14]:
 # Transform results into DataFrame structure
results_1 = pd.DataFrame(rules)

In [15]:
results_1

items   support  \
0     (HP 63XL Ink, 5pack Nylon Braided USB C cables)  0.005733   
1   (Anker 2-in-1 USB Card Reader, FEIYOLD Blue li...  0.007999   
2   (Falcon Dust Off Compressed Gas, Apple Lightni...  0.004533   
3   (AutoFocus 1080p Webcam, SanDisk Ultra 64GB card)  0.005333   
4                       (iPhone 11 case, HP 63XL Ink)  0.005866   
..                                                ...       ...   
79  (Dust-Off Compressed Gas 2 pack, 0, HP 61 ink,...  0.003866   
80  (Dust-Off Compressed Gas 2 pack, 0, HP 61 ink,...  0.003999   
81  (SanDisk Ultra 64GB card, Dust-Off Compressed ...  0.003733   
82  (SanDisk Ultra 64GB card, Dust-Off Compressed ...  0.004399   
83  (Dust-Off Compressed Gas 2 pack, 0, VIVO Dual ...  0.004533   

                                   ordered_statistics  
0   [((5pack Nylon Braided USB C cables), (HP 63XL...  
1   [((Anker 2-in-1 USB Card Reader), (FEIYOLD Blu...  
2   [((Apple Lightning to USB cable), (Falcon Dust...  
3   [((AutoFocus 1080p Webcam), (SanDisk Ultra 64G...  
4   [((iPhone 11 case), (HP 63XL Ink), 0.372881355...  
..                                                ...  
79  [((HP 61 ink, FEIYOLD Blue light Blocking Glas...  
80  [((HP 61 ink, Dust-Off Compressed Gas 2 pack, ...  
81  [((SanDisk Ultra 64GB card, Nylon Braided Ligh...  
82  [((SanDisk Ultra 64GB card, Nylon Braided Ligh...  
83  [((Dust-Off Compressed Gas 2 pack, VIVO Dual L...  

[84 rows x 3 columns]

In [16]:
supportz = results_1.support

In [17]:

one_val = []
two_val = []
three_val = []
four_val= []

In [18]:
# Establish a for loop to cycle through a list.
for i in range(results_1.shape[0]):
 listz = results_1['ordered_statistics'][i][0]
 one_val.append(list(listz[0]))
 two_val.append(list(listz[1]))
 three_val.append(listz[2])
 four_val.append(listz[3])

In [19]:

lhs = pd.DataFrame(one_val)
rhs = pd.DataFrame(two_val)
confidence = pd.DataFrame(three_val, columns=['confidence'])
lift = pd.DataFrame(four_val, columns=['lift'])

In [20]:

finalz= pd.concat([lhs, rhs, supportz, confidence, lift], axis=1)
finalz.fillna(value=' ', inplace=True)

In [21]:

finalz


0                                     1  \
0   5pack Nylon Braided USB C cables                                         
1       Anker 2-in-1 USB Card Reader                                         
2       Apple Lightning to USB cable                                         
3             AutoFocus 1080p Webcam                                         
4                     iPhone 11 case                                         
..                               ...                                   ...   
79                         HP 61 ink   FEIYOLD Blue light Blocking Glasses   
80                         HP 61 ink        Dust-Off Compressed Gas 2 pack   
81           SanDisk Ultra 64GB card  Nylon Braided Lightning to USB cable   
82           SanDisk Ultra 64GB card  Nylon Braided Lightning to USB cable   
83    Dust-Off Compressed Gas 2 pack      VIVO Dual LCD Monitor Desk mount   

                                       2                                    0  \
0                                                                 HP 63XL Ink   
1                                         FEIYOLD Blue light Blocking Glasses   
2                                              Falcon Dust Off Compressed Gas   
3                                                     SanDisk Ultra 64GB card   
4                                                                 HP 63XL Ink   
..                                   ...                                  ...   
79                                                                          0   
80  Nylon Braided Lightning to USB cable                                    0   
81                                                                          0   
82                                                                          0   
83         Screen Mom Screen Cleaner kit                                    0   

                                       1                                 2  \
0                                                                            
1                                                                            
2                                                                            
3                                                                            
4                                                                            
..                                   ...                               ...   
79        Dust-Off Compressed Gas 2 pack  VIVO Dual LCD Monitor Desk mount   
80         Screen Mom Screen Cleaner kit                                     
81        Dust-Off Compressed Gas 2 pack     Screen Mom Screen Cleaner kit   
82        Dust-Off Compressed Gas 2 pack  VIVO Dual LCD Monitor Desk mount   
83  Nylon Braided Lightning to USB cable                                     

     support  confidence      lift  
0   0.005733    0.300699  3.790833  
1   0.007999    0.271493  4.122410  
2   0.004533    0.290598  4.843951  
3   0.005333    0.377358  3.840659  
4   0.005866    0.372881  4.700812  
..       ...         ...       ...  
79  0.003866    0.235772  3.947608  
80  0.003999    0.410959  3.171402  
81  0.003733    0.220472  4.593788  
82  0.004399    0.259843  4.350622  
83  0.004533    0.288136  3.022804  

[84 rows x 9 columns]

C3. Association Rules Table:

In [22]:
finalz.columns = ['lhs', 1, 2, 'rhs', 1, 2, 'support', 'confidence', 'lift']
finalzz = finalz[['lhs', 'rhs', 'support', 'confidence', 'lift']]
finalzz

lhs                                  rhs  \
0   5pack Nylon Braided USB C cables                          HP 63XL Ink   
1       Anker 2-in-1 USB Card Reader  FEIYOLD Blue light Blocking Glasses   
2       Apple Lightning to USB cable       Falcon Dust Off Compressed Gas   
3             AutoFocus 1080p Webcam              SanDisk Ultra 64GB card   
4                     iPhone 11 case                          HP 63XL Ink   
..                               ...                                  ...   
79                         HP 61 ink                                    0   
80                         HP 61 ink                                    0   
81           SanDisk Ultra 64GB card                                    0   
82           SanDisk Ultra 64GB card                                    0   
83    Dust-Off Compressed Gas 2 pack                                    0   

     support  confidence      lift  
0   0.005733    0.300699  3.790833  
1   0.007999    0.271493  4.122410  
2   0.004533    0.290598  4.843951  
3   0.005333    0.377358  3.840659  
4   0.005866    0.372881  4.700812  
..       ...         ...       ...  
79  0.003866    0.235772  3.947608  
80  0.003999    0.410959  3.171402  
81  0.003733    0.220472  4.593788  
82  0.004399    0.259843  4.350622  
83  0.004533    0.288136  3.022804  

[84 rows x 5 columns]

We can demonstrate mathematically that "5pack Nylon Braided USB C cables" and "HP 63XL Ink" have the highest combination of values for our three metrics (support, confidence, lift) after running the final findings to build the association rules table:
* As seen above, support equates to 0.006, confidence = 0.301 and lift equates to around 3.80


C4. Top Three Rules: 

1. This rule, which we are confident in, shows that 30% of customers who bought the Nylon 5 pack cables also bought the HP 63XL Ink. With a support score of 0.0057, the most straightforward measure of support shows that slightly more than 5% of purchases include both goods. The lift value of 3.8 indicates that a client is 3.8 times more likely to acquire the "HP 63XL Ink" after first purchasing the "5pack Nylon Braided USB C cords."

2. Based on the table above, we are confident that 27% of customers who bought "Anker 2-in-1 USB Card Reader" followed that up with "SanDisk Ultra 64GB card" together. The lift value turned out to be 4.12, 4.12 times more likely to purchase after antecedent purchase. Support = 0.00799 (7.9% of all transcations contained both items)

3. We are confident that 29% of customers who bought "Apple Lightening to USB Cable" also bought "Falcon Dust-Off Compressed Gas". 0.0045 for support and Lift equates to 4.84. 4.84 times more likely to acquire subsequent after antecedent purchase. 

D1: Significance of SLC (Support/Lift/Confidence)

An examination of the market basket might help comprehend consumer purchasing patterns. Support gauges how frequently a dataset's association rules occur. An association rule's strength is measured by its confidence. In market basket analysis, this is how likely a second product will be in the basket if the first is. To determine it, divide the number of transactions that include only the first product (or group of effects), the antecedent component of the association rule, by the number of transactions that include both products (or group of products). Lift evaluates how well an association rule—a targeting model—predicts a particular outcome compared to a random selection. Lift is, therefore, the proportion of the goal response to the average response.

Our highest confidence level is in rule #3, which is at around 38 percent, while the #1 rule is only at 30 percent (based on its examination in conjunction with our other metrics of relevance).
Again, there is not enough evidence to be convincing to justify pairing any of the top three rules because it does not happen in more than 5% of all transactions.
Finally, the lift ratio gives us hope that a client will buy the following item after purchasing the antecedent. The correlation between buying an "Apple Lightning to USB cable" and then purchasing some "Falcon Dust-Off Compressed Gas" demonstrates our highest lift metric at "4.84 times more likely".

D2. Practical Significance of Findings 


Since we can't truly be sure that any collection of things will be bought even half the time, we don't think these results have a lot of practical application. We have a better probability of correctly guessing the result of a coin toss, don't we? We can see that the likelihood that a consumer will choose to buy the consequent is around four times higher if one of the precursors is chosen for acquisition. 
Therefore, if a person purchases, for instance, a Nylon USB cables, they are almost four times more likely to purchase HP ink for the printer.
We really don't have a lot to go on with these results. Before making any conclusions about relevance, perhaps we should compile more data. Of course, additional research is advised.

D3. Plan of Action

Based on the past analysis and discussion of relevance, we thus advise the company's decision-makers not carrying out with the plan of advertising our service by offering discounted items in return for subscribing to our broadband services. No pairings showed that customers who used telecom services would want or require some related item, and our market basket analysis of this transaction information did not reveal any importance either.

This means that we might advise one of those products for a prospective customer discount and a marketing offer if we had discovered a substantial association with, for example, many transactions where consumers acquired two related telecoms peripherals. We didn't discover that. We discovered ink being purchased at the same time as a webcam and an ethernet cable, suggesting that we might be looking for a relationship.
At this time, no action is necessary. Before our data science team can recommend confident action, more data needs to be collected and processed.